include/tscpp/util/Hash.h

```diff
@@ -0,0 +1,153 @@
+/** @file Basic hash function support.
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+#include <cctype>
+#include "tscpp/util/MemSpan.h"
+#include "tscpp/util/TextView.h"
+
+namespace ts
+{
+/** Base protocol class for hash functors.
+ *
+ * Each specific hash function embedded in a hash functor is a subclass of this class and
+ * follows this API. Subclasses should override the return type to return the subclass type.
+ *
+ * The main purpose of this is to allow run time changes in hashing, which is required in various
+ * circumstances.
+ */
+struct HashFunctor {
+  using self_type = HashFunctor; ///< Self reference type.
+
+  /// Pass @a data to the hashing function.
+  virtual self_type &update(std::string_view const &data) = 0;
+
+  /// Finalize the hash function output.
+  virtual self_type & final() = 0;
```

Comment:
Interesting.  Although final is a keyword, it's not a reserve word.  Consistency is the hobgoblin of little programming languages.

Reply:
We could change that, it's a legacy from pre-eleventy days.